File: app/types/projects.ts
export type Project = {
    project_id: string;
    name: string;
    frontend: string;
    backend: string;
    db: string;
    user_id: string;
    container_id: string;
    status: string;
    last_online: string;
    roots: ProjectRoots;
}

export type ProjectRoots = {
    frontend_root: string | null;
    backend_root: string | null;
    db_root: string | null;
}

export type GithubRepo = {
    id: number;
    name: string;
    full_name: string;
    private: boolean;
    url: string;
    default_branch: string;
    updated_at: string;
}

export interface ProjectEnv {
  key: string;
  value: string;
  is_secret: boolean;
}

export interface ProjectDbColumn {
  column: string;
  type: string;
  nullable: boolean;
}

export interface ProjectPage {
  route: string;
  file: string; // relative to /app/workspace/frontend/{name}/
}

export interface ProjectEndpoint {
  method: string;
  path: string;
  file: string; // relative to /app/workspace/backend/{name}/
  handler: string;
}

export interface ProjectGroupFileMeta {
  type?: "wrapper" | "hook" | "helper" | "middleware";
  category?: "http" | "validation" | "auth" | "payment";
  compatibility?: string;
  library?: string;
  style?: string;
  colourScheme?: string;
  [key: string]: unknown;
}

export interface ProjectGroupFile {
  name: string;
  filepath: string; // relative to root
  meta?: ProjectGroupFileMeta;
}

export interface ProjectGroup {
  name: string;
  filepath: string;
  type: "folder" | "file";
  context: "frontend" | "backend";
  meta: ProjectGroupFileMeta;
  children: ProjectGroup[];
}

export type ProjectMetaData = {
  envs: ProjectEnv[];
  db_schema: Record<string, ProjectDbColumn[]>;
  pages: ProjectPage[];
  endpoints: ProjectEndpoint[];
  groups: ProjectGroup[];
  updated_at: string | null;
}

export interface ServiceStatus {
  frontend: boolean;
  backend: boolean;
  database: boolean;
  container: boolean;
}

File: backend/structs/projects.go
package structs

import "time"

type Service struct {
	Name      string `json:"name"`
	Framework string `json:"framework"`
}

type Project struct {
	ID          string     `json:"project_id"`
	Name        string     `json:"name"`
	Status      string     `json:"status"`
	ContainerID string     `json:"container_id"`
	CreatedAt   time.Time  `json:"created_at"`
	LastOnline  *time.Time `json:"last_online"`
	Services    []Service  `json:"services"`
}

type CreateProjectRequest struct {
	Name      string `json:"name"`
	Backend   string `json:"backend"`
	Frontend  string `json:"frontend"`
	Database  string `json:"db"`
	ImportURL string `json:"import_url"`
	Envs      []Env  `json:"envs"`
}

type Env struct {
	Key      string `json:"key"`
	Value    string `json:"value"`
	IsSecret bool   `json:"is_secret"`
}

type Page struct {
	Route string `json:"route"`
	File  string `json:"file"`
}

type Endpoint struct {
	Method string `json:"method"`
	Path   string `json:"path"`
	File   string `json:"file"`
}

type ProjectGroup struct {
	Name     string            `json:"name"`
	Type     string            `json:"type"`
	Context  string            `json:"context,omitempty"`
	Filepath string            `json:"filepath"`
	Meta     map[string]string `json:"meta,omitempty"`
	Children []ProjectGroup    `json:"children,omitempty"`
}

type ScaffoldResult struct {
	Pages     []Page     `json:"pages"`
	Endpoints []Endpoint `json:"endpoints"`
}

type ScanResult struct {
	FrontendRoot      string
	BackendRoot       string
	DBRoot            string
	FrontendFramework string
	BackendFramework  string
	DBFramework       string
	Pages             []Page
	Endpoints         []Endpoint
	FrontendGroups    []ProjectGroup
	BackendGroups     []ProjectGroup
}
